burn-train/src/learner/base.rs

```diff
@@ -1,91 +1,67 @@
-use burn_core::lr_scheduler::LRScheduler;
-use burn_core::module::{ADModule, Module};
+use crate::components::LearnerComponents;
+use burn_core::lr_scheduler::LrScheduler;
+use burn_core::module::Module;
 use burn_core::optim::Optimizer;
-use burn_core::tensor::backend::ADBackend;
+use burn_core::tensor::backend::Backend;
 use std::sync::atomic::{AtomicBool, Ordering};
 use std::sync::Arc;
 
 /// Learner struct encapsulating all components necessary to train a Neural Network model.
 ///
 /// To create a learner, use the [builder](crate::learner::LearnerBuilder) struct.
-pub struct Learner<B, Model, Optim, LR, TrainOutput, ValidOutput>
-where
-    B: ADBackend,
-    LR: LRScheduler,
-    Model: ADModule<B>,
-    Optim: Optimizer<Model, B>,
-{
-    pub(super) model: Model,
-    pub(super) optim: Optim,
-    pub(super) lr_scheduler: LR,
-    pub(super) num_epochs: usize,
-    pub(super) callback: Box<dyn LearnerCallback<TrainOutput, ValidOutput>>,
-    pub(super) checkpoint: Option<usize>,
-    pub(super) checkpointer_model: CheckpointModel<Model, B>,
-    pub(super) checkpointer_optimizer: CheckpointOptim<Optim, Model, B>,
-    pub(super) checkpointer_scheduler: CheckpointScheduler<LR>,
-    pub(super) grad_accumulation: Option<usize>,
-    pub(super) devices: Vec<B::Device>,
-    pub(super) interrupter: TrainingInterrupter,
+pub struct Learner<C: LearnerComponents> {
+    pub(crate) model: C::Model,
+    pub(crate) optim: C::Optimizer,
+    pub(crate) lr_scheduler: C::LrScheduler,
+    pub(crate) num_epochs: usize,
+    pub(crate) checkpoint: Option<usize>,
+    pub(crate) grad_accumulation: Option<usize>,
+    pub(crate) checkpointer: Option<LearnerCheckpointer<C>>,
+    pub(crate) devices: Vec<<C::Backend as Backend>::Device>,
+    pub(crate) callback: C::Callback,
+    pub(crate) interrupter: TrainingInterrupter,
 }
 
-type CheckpointModel<M, B> = Option<Box<dyn Checkpointer<<M as Module<B>>::Record>>>;
-type CheckpointOptim<O, M, B> = Option<Box<dyn Checkpointer<<O as Optimizer<M, B>>::Record>>>;
-type CheckpointScheduler<LR> = Option<Box<dyn Checkpointer<<LR as LRScheduler>::Record>>>;
+#[derive(new)]
+pub(crate) struct LearnerCheckpointer<C: LearnerComponents> {
+    model: C::CheckpointerModel,
+    optim: C::CheckpointerOptimizer,
+    lr_scheduler: C::CheckpointerLrScheduler,
+}
 
-impl<B, Model, Optim, LR, TrainOutput, ValidOutput>
-    Learner<B, Model, Optim, LR, TrainOutput, ValidOutput>
-where
-    ValidOutput: Send + Sync + 'static,
-    TrainOutput: Send + Sync + 'static,
-    B: ADBackend,
-    Model: ADModule<B>,
-    Optim: Optimizer<Model, B>,
-    LR: LRScheduler,
-{
-    pub(super) fn checkpoint(
-        model: &Model,
-        optim: &Optim,
-        scheduler: &LR,
-        checkpointer_model: &CheckpointModel<Model, B>,
-        checkpointer_optimizer: &CheckpointOptim<Optim, Model, B>,
-        checkpointer_scheduler: &CheckpointScheduler<LR>,
+impl<C: LearnerComponents> LearnerCheckpointer<C> {
```

Comment:
I'm starting to think there is too many times the letter C: callback, checkpoint, components

Reply:
We could use LC ?